build_dependency_tree.py

```diff
@@ -3,76 +3,202 @@
 import sys
 import re
 from pathlib import Path
-from typing import Optional
-from typing import List
+from typing import Optional, Union
+from typing import List, Dict
 
 
 def extract_files_from_directory():
     """FUnction recurses through the directory to find all file-like objects"""
 
+    def is_cpp_file(file: Union[Path, str]):
+        file_types = ('.c', '.cpp', '.cxx', '.C', '.h', '.hpp', '.hxx')
+        return_value = False
+        for file_type in file_types:
+            return_value = return_value or file_type in str(file)
+            if return_value:
+                return True
+        return False
+
     def extract_files(dir_or_file: Path, list_o_files: List[Optional[Path]]):
         if dir_or_file.is_dir():
+            print(f'In dir: {str(dir_or_file.absolute())}')
             for path in dir_or_file.glob("*"):
                 extract_files(path, list_o_files)
         else:
-            list_o_files.append(dir_or_file)
+            if is_cpp_file(dir_or_file.name):
+                print(f'Appending file: {str(dir_or_file.absolute())}')
+                list_o_files.extend([dir_or_file])
+        return list_o_files
+
+    def flatten_list(lists: Optional[List[List]]):
+        return [
+                item for sublist in lists for item in sublist
+                if len(sublist) > 0]
 
-    return [extract_files(Path(arg), []) for arg in sys.argv]
+    return flatten_list(
+            [extract_files(Path(arg), []) for arg in sys.argv])
 
 
+# I think we can get rid of thise function
 def separate_headers_and_implementations(files_in_dir: List[Path]):
-    """ 
-    Separates files from current directory into header and implementation files.
-    
-    :param files_in_dir: List of paths to files C++ files in current directory
-    
-    return: returns: tuple of list containing paths to the header files in 
-    first slot and implementation files in the second
+    """
+    Separates files from current directory into header and implementation files
+
+    :param files_in_dir:    List of paths to files C++ files in current
+                            directory
+    return: returns:        tuple of list containing paths to the header files
+                            in first slot and implementation files in the
+                            second
     """
     headers = [file for file in files_in_dir if file.suffix in ('h', 'hpp')]
     impls = [file for file in files_in_dir if file.suffix in ('c', 'cpp')]
     return headers, impls
 
 
-def create_single_file_dependency_list(file: Path):
+# NOTE: Do we want to make re.Pattern object a function parameter?
+# KI: No, but also don't want it to be a global variable. This screams classes
+#
+# match either the pattern `<...>` or `"..."`
+include_statement_pattern = re.compile(r"(<.+?>)|(\".+?\")")
```

Comment:
```suggestion
include_statement_pattern = re.compile(r"(<.+?>|\".+?\")")
```

Reply:
The parentheses are their to add groups. It may be important down the line to distinguish formatting of graph output and whether a header is included as <> or "" header file